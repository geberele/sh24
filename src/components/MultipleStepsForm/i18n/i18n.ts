import { I18N } from './i18n.types'

export const en: I18N = {
  question1: 'What is your name?',
  question2: 'What is your email address?',
  question3: 'What service are you here for?',
  options1: 'STI Testing',
  options2: 'Contraception',
  options3: 'Other',
  labelName: 'Name',
  labelEmail: 'Email address',
  previewTitle: 'Preview',
  prev: 'Back',
  next: 'Next',
  errorMessage1: 'Please insert a valid name (2-25 characters)',
  errorMessage2: 'Please insert a valid email',
  errorMessage3: 'Please select an option',
}

export const it: I18N = {
  question1: 'Come ti chiami?',
  question2: 'Quale Ã¨ la tua email?',
  question3: 'Che tipo di servizio ti interessa?',
  options1: 'Test STI',
  options2: 'Contraccettivo',
  options3: 'Altro',
  labelName: 'Nome',
  labelEmail: 'Email',
  previewTitle: 'Preview',
  prev: 'Indietro',
  next: 'Avanti',
  errorMessage1: 'Inserisci un nome valido (2-25 lettere)',
  errorMessage2: 'Inserisci un\'email valida',
  errorMessage3: 'Seleziona un\'opzione per favore',
}
