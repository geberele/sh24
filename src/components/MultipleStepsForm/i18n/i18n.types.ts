export interface I18N {
  question1: string
  question2: string
  question3: string
  options1: string
  options2: string
  options3: string
  labelName: string
  labelEmail: string
  previewTitle: string
  prev: string
  next: string
  errorMessage1: string
  errorMessage2: string
  errorMessage3: string
}
