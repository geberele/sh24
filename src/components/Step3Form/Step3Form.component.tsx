import React from 'react'
import { useForm } from 'react-hook-form'
import cn from 'classnames'

export interface Step3FormProps {
  question: string
  next: string
  prev: string
  value: string
  errorMessage: string
  options: Options[]
  nextStep: () => void
  prevStep: () => void
  onChange: (value: string) => void
}

export interface Options {
  label: string
  value: string
}

export const Step3Form = ({
  question,
  next,
  prev,
  options,
  nextStep,
  prevStep,
  onChange,
  value,
  errorMessage,
}: Step3FormProps) => {
  const { register, handleSubmit, errors } = useForm()

  return (
    <form onSubmit={handleSubmit(nextStep)}>
      <h1>{question}</h1>
      <div className="radio-buttons">
        {options.map((option, key) => (
          <div
            key={key}
            className={cn('radio-button', { 'selected': value === option.value })}
          >
            <label htmlFor={option.value}>{option.label}</label>
            <input
              type="radio"
              name="service"
              value={option.value}
              id={option.value}
              onChange={e => onChange(e.target.value)}
              checked={value === option.value}
              ref={register({ required: true })}
              data-testid="input-service"
              className={cn('hidden', { 'checked': value === option.value })}
            />
          </div>
        ))}
      </div>

      {errors.service &&
        <div
          className="error"
          data-testid="input-error"
        >
          {errorMessage}
        </div>
      }

      <div className="navigation">
        <button
          onClick={prevStep}
          className="navigationLeft"
          data-testid="button-prev"
        >
          {prev}
        </button>
        <button
          type="submit"
          className="navigationRight"
          data-testid="button-next"
        >
          {next}
        </button>
      </div>
    </form>
  )
}
