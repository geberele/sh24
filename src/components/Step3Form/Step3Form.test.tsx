import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { fireEvent, waitFor, screen } from '@testing-library/dom'
import { Options, Step3Form, Step3FormProps } from './Step3Form.component'
import { en } from '../MultipleStepsForm/i18n/i18n'

describe('Step3Form', () => {
  const prevStep = jest.fn()
  const nextStep = jest.fn()

  const options: Options[] = [
    { label: en.options1, value: 'one' },
    { label: en.options2, value: 'two' },
    { label: en.options3, value: 'three' },
  ]

  const props: Step3FormProps = {
    question: en.question2,
    options: options,
    prev: en.prev,
    next: en.next,
    value: '',
    prevStep: prevStep,
    nextStep: nextStep,
    onChange: () => {
      return
    },
    errorMessage: en.errorMessage3,
  }

  const setup = () => {
    const utils = render(<Step3Form {...props}/>)
    const radios = screen.getAllByTestId('input-service') as HTMLInputElement[]
    const prevButton = screen.getByTestId('button-prev') as HTMLInputElement
    const submit = screen.getByTestId('button-next') as HTMLInputElement
    return {
      radios,
      prevButton,
      submit,
      ...utils,
    }
  }

  it('renders', () => {
    const { asFragment } = setup()
    expect(asFragment()).toMatchSnapshot()
  })

  it('returns error if no option is selected', async () => {
    const { submit } = setup()
    fireEvent.click(submit)
    const error = await waitFor(() => screen.getByTestId('input-error'))
    expect(error.textContent).toBe(en.errorMessage3)
  })

  it('submits the form if email is valid', async () => {
    const { radios, rerender, submit } = setup()
    rerender(<Step3Form {...{ ...props, value: 'one' }} />)
    await waitFor(() => expect(radios[0].classList.contains('checked')).toEqual(true))
    fireEvent.click(submit)
    await waitFor(() => expect(nextStep).toHaveBeenCalledTimes(1))
  })

  it('goes to the previous step', async () => {
    const { prevButton } = setup()
    fireEvent.click(prevButton)
    await waitFor(() => expect(prevStep).toHaveBeenCalledTimes(1))
  })
})
