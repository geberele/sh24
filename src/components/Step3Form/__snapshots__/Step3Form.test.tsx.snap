// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step3Form renders 1`] = `
<DocumentFragment>
  <form>
    <h1>
      What is your email address?
    </h1>
    <div
      class="radio-buttons"
    >
      <div
        class="radio-button"
      >
        <label
          for="one"
        >
          STI Testing
        </label>
        <input
          class="hidden"
          data-testid="input-service"
          id="one"
          name="service"
          type="radio"
          value="one"
        />
      </div>
      <div
        class="radio-button"
      >
        <label
          for="two"
        >
          Contraception
        </label>
        <input
          class="hidden"
          data-testid="input-service"
          id="two"
          name="service"
          type="radio"
          value="two"
        />
      </div>
      <div
        class="radio-button"
      >
        <label
          for="three"
        >
          Other
        </label>
        <input
          class="hidden"
          data-testid="input-service"
          id="three"
          name="service"
          type="radio"
          value="three"
        />
      </div>
    </div>
    <div
      class="navigation"
    >
      <button
        class="navigationLeft"
        data-testid="button-prev"
      >
        Back
      </button>
      <button
        class="navigationRight"
        data-testid="button-next"
        type="submit"
      >
        Next
      </button>
    </div>
  </form>
</DocumentFragment>
`;
